mmv1/third_party/terraform/services/container/resource_container_cluster.go.tmpl

```diff
@@ -1763,38 +1879,42 @@ func ResourceContainerCluster() *schema.Resource {
 						"peering_name": {
 							Type:         schema.TypeString,
 							Computed:     true,
-							Description: `The name of the peering between this cluster and the Google owned VPC.`,
+							Description:  `The name of the peering between this cluster and the Google owned VPC.`,
 						},
 						"private_endpoint": {
 							Type:         schema.TypeString,
 							Computed:     true,
-							Description: `The internal IP address of this cluster's master endpoint.`,
+							Description:  `The internal IP address of this cluster's master endpoint.`,
 						},
 						"private_endpoint_subnetwork": {
-							Type:         schema.TypeString,
-							Optional:     true,
-							ForceNew:     true,
-							AtLeastOneOf: privateClusterConfigKeys,
+							Type:             schema.TypeString,
+							Optional:         true,
+							Deprecated:       "use control_plane_endpoints_config.ip_endpoints_config.private_endpoint_subnetwork instead",
+							ConflictsWith:    []string{"control_plane_endpoints_config.0.ip_endpoints_config.0.private_endpoint_subnetwork"},
+							ForceNew:         true,
+							AtLeastOneOf:     privateClusterConfigKeys,
 							DiffSuppressFunc: containerClusterPrivateClusterConfigSuppress,
-							Description: `Subnetwork in cluster's network where master's endpoint will be provisioned.`,
+							Description:      `Subnetwork in cluster's network where master's endpoint will be provisioned.`,
 						},
 						"public_endpoint": {
 							Type:         schema.TypeString,
 							Computed:     true,
 							Description: `The external IP address of this cluster's master endpoint.`,
 						},
 						"master_global_access_config": {
-							Type:         schema.TypeList,
-							MaxItems:     1,
-							Optional:     true,
-							Computed:     true,
-							AtLeastOneOf: privateClusterConfigKeys,
-							Description: "Controls cluster master global access settings.",
+							Type:          schema.TypeList,
+							MaxItems:      1,
+							Optional:      true,
+							Deprecated:    "use control_plane_endpoints_config.ip_endpoints_config.global_access instead",
+							ConflictsWith: []string{"control_plane_endpoints_config.0.ip_endpoints_config.0.global_access"},
+							Computed:      true,
+							AtLeastOneOf:  privateClusterConfigKeys,
+							Description:   "Controls cluster master global access settings.",
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
 									"enabled": {
 										Type:        schema.TypeBool,
-										Required:    true,
+										Optional:    true,
```

Comment:
Can you explain the situation a user would want to set `master_global_access_config {}` in configuration? In case you are unaware, `Required` only applies in the scope of the parent block, so this specifically presents `master_global_access_config {}` from being specified, but does not force specifying a value when it is not present.

Reply:
Similarly, we may want to consider making subfields of some new fields `Required: true`, unless it is valid to specify an empty block.

Reply:
This was just to match the underlying API, which was built before proto3 added the optional keyword back. So we only used this to get presence information for the enabled bool, so we could send private_cluster_config patches without having to do a full read-modify-write loop. It's unnecessary here, so I'll just go back to using Required. I only did this to make sure we can do the migration; I didn't grok the scoping.

Reply:
Oh man, that's a whole yak shave on it's own Proto3 optional is a slightly different thing than optional. That's equivalent to turning a bool into a *bool, which we don't have the types to express (and would be breaking anyways, as it would for an API to retrofit as it changes the meaning of the null value).